### PR TITLE
R1:  Free5GC SMF 

### PR DESCRIPTION
**Pull request by n2vo**
Steps used to build/create/test:
- cd ./free5gc-operator
- make generate
- make manifests
- make install
- kubectl apply -f ./config/crd/bases/workload.nephio.org_smfdeployments.yaml
- make docker-build
- make docker-push
- make deploy
- k apply -f test/smfdeploy.yaml
- k apply -f test/upfdeploy.yaml 

Checking:
- k get pods -A

Sample local K8s dev environment 
- Ubuntu Server 20.04.5/amd64
- go1.20.4 linux/amd64
- kubeadm, kubelet, kubectl 1.27.1
-  2 nodes:  c3-cp1 & c3-node1

Test Results:

avo@c4-cp1:~/dev/n2vo/free5gc/free5gc-operator$ k get pods -A
NAMESPACE                 NAME                                                   READY   STATUS    RESTARTS   AGE
free5gc-operator-system   free5gc-operator-controller-manager-6fb9cf7646-r57gt   2/2     Running   0          50s
kube-flannel              kube-flannel-ds-5dvqx                                  1/1     Running   0          8d
kube-flannel              kube-flannel-ds-85b8q                                  1/1     Running   0          8d
kube-system               coredns-5d78c9869d-8m9fr                               1/1     Running   0          8d
kube-system               coredns-5d78c9869d-gtjvm                               1/1     Running   0          8d
kube-system               etcd-c4-cp1                                            1/1     Running   0          8d
kube-system               kube-apiserver-c4-cp1                                  1/1     Running   0          8d
kube-system               kube-controller-manager-c4-cp1                         1/1     Running   0          8d
kube-system               kube-multus-ds-n9krs                                   1/1     Running   0          8d
kube-system               kube-multus-ds-v5fj7                                   1/1     Running   0          8d
kube-system               kube-proxy-bxqm4                                       1/1     Running   0          8d
kube-system               kube-proxy-wx9r5                                       1/1     Running   0          8d
kube-system               kube-scheduler-c4-cp1                                  1/1     Running   0          8d
smf-1                     free5gc-smf-1-748b45f749-q5w89                         1/1     Running   0          31s
upf-1                     free5gc-upf-1-668c6769bd-wkvk5                         1/1     Running   0          23s

### CODE REVIEW

File: free5gc-operator/Dockerfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ RUN go mod download
 
 # Copy the go source
 COPY main.go main.go
-COPY api/ api/
+#COPY api/ api/
```
Comment (electrocucaracha) [inline free5gc-operator/Dockerfile:16]:
```suggestion
```

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+// SMFDeploymentReconciler reconciles a SMFDeployment object
+type SMFDeploymentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+type SMFcfgStruct struct {
+	PFCP_IP string // N4
+	// N4_IP   string
+	// N11_IP  string
```
Comment (electrocucaracha) [inline free5gc-operator/controllers/smfdeployment_controller.go:56]:
```suggestion
```

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+type SMFAnnotation struct {
+	Name      string `json:"name"`
+	Interface string `json:"interface"`
+	IP        string `json:"ip"`
+	Gateway   string `json:"gateway"`
+}
+
+// Check string -> int/string -uint16 conversion
+func getSMFResourceParams(smfSpec workloadv1alpha1.SMFDeploymentSpec) (int32, *apiv1.ResourceRequirements, error) {
+	// Placeholder for Capacity calculation. Reqiurce requirements houlw be calculated based on DL, UL.
```
Comment (electrocucaracha) [inline free5gc-operator/controllers/smfdeployment_controller.go:68]:
```suggestion
	// Placeholder for Capacity calculation. Require requirements whould be calculated based on DL, UL.
```

File: free5gc-operator/test/nad/smf-n4-nad.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+---
+# Source: free5gc/charts/free5gc-upf/templates/upf-n4-nad.yaml
+#
+# Software Name : towards5gs-helm
+# SPDX-FileCopyrightText: Copyright (c) 2021 Orange
```
Comment (electrocucaracha) [inline free5gc-operator/test/nad/smf-n4-nad.yaml:5]:
```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2023 The Nephio Authors.
```

File: free5gc-operator/controllers/smfdeployment_controller.go (hunk 3 above)
Comment (denysaleksandrov) [reply free5gc-operator/controllers/smfdeployment_controller.go:68]: This comment should be deleted. copy-paste from UPFDeployment.

**Comment by vireshnavalli** [comment]
Hi  @tliron , @denysaleksandrov 
Please correct me if I am wrong, Ideally operator should watch for crds and automatically reconcile it and spawn the upf/smf/amd pod correct?

The steps mentioned in the description by @n2vo , for me it looks like we are explicitly applying the CRDs.

**Comment by denysaleksandrov** [comment]
@vireshnavalli correct, SMF controller starts and wait for SMFDeployment kinds to be created/applied to a cluster.
But before a controller starts we need to register SMFDeployment kind with a cluster by applying a CRD definition found [here](https://github.com/nephio-project/api/blob/main/config/crd/bases/workload.nephio.org_smfdeployments.yaml). Otherwise we get such an error:
```
...    ERROR   setup   unable to create controller     {"controller": "SMFDeployment", "error": "no kind is registered for the type v1alpha1.SMFDeployment in scheme \"pkg/runtime/scheme.go:100\""}
```

The following line in the description was used during development and testing for the same purpose:
```
...
kubectl apply -f ./config/crd/bases/workload.nephio.org_smfdeployments.yaml
...
```

**Comment by denysaleksandrov** [comment]
@vireshnavalli, Correct. Local CRDs in this repo are used only for testing and development. I  would personally keep them here as well for that purpose.

File: free5gc-operator/controllers/free5gc_smfcfg.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    addr: {{ .PFCP_IP }}
+  smfName: SMF
+  snssaiInfos:
+    - sNssai:
+        sst: 1
+        sd: 010203
+      dnnInfos:
+        - dnn: internet
+          dns:
+            ipv4: 8.8.8.8
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/free5gc_smfcfg.go:34]: This should probably be configurable, but that come come in a later PR (non-blocking comment).

File: free5gc-operator/controllers/free5gc_smfcfg.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    - sNssai:
+        sst: 2
+        sd: 112234
+      dnnInfos:
+        - dnn: internet
+          dns:
+            ipv4: 8.8.8.8
+  plmnList:
+    - mcc: "208"
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/free5gc_smfcfg.go:50]: In our discussions, we talked about a "3gpp config" resource that would contain this information. Shouldn't this be in the SMFDeployment resource?

File: free5gc-operator/controllers/free5gc_smfcfg.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      mnc: "93"
+  userplaneInformation:
+    upNodes:
+      gNB1:
+        type: AN
+      UPF:
+        type: UPF
+        nodeID: 10.100.50.241
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/free5gc_smfcfg.go:58]: This shouldn't be hardcoded either.

File: free5gc-operator/controllers/free5gc_smfcfg.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        sNssaiUpfInfos:
+          - sNssai:
+              sst: 1
+              sd: 010203
+            dnnUpfInfoList:
+              - dnn: internet
+                pools:
+                  - cidr: 10.1.0.0/17
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/free5gc_smfcfg.go:66]: ditto this and the CIDR below

These should be coming as inputs in the SMFDeployment, shouldn't they?

File: free5gc-operator/controllers/free5gc_smfuerouting.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package controllers
+
+var Uerouting string = `
+info:
+  version: 1.0.1
+  description: Routing information for UE
+ueRoutingInfo:
+  UE1:
+    members:
+    - imsi-208930000000003
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/free5gc_smfuerouting.go:10]: I don't know what all these things mean, but they look like things we should be constructing based on the SMFDeployment.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+	// TODO: increase number of replicas based on NFDeployment.Capacity.MaxSessions
+	var replicas int32 = 1
+	var cpuLimit string
+	var cpuRequest string
+	var memoryLimit string
+	var memoryRequest string
+
+	// spec.Capacity.MaxSessions <1000 -> 128Mi
+	// spec.Capacity.MaxNFConnections <10 -> 128Mi
+	if smfSpec.Capacity.MaxSessions < 1000 && smfSpec.Capacity.MaxNFConnections < 10 {
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:79]: Looks good. Glad to see this here. Though, IMO this is the sort of thing that belongs in a kpt function not an operator. But that is subject to debate, it's fine to do it here.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+	resources.Requests = make(apiv1.ResourceList)
+	resources.Requests[apiv1.ResourceCPU] = resource.MustParse(cpuRequest)
+	resources.Requests[apiv1.ResourceMemory] = resource.MustParse(memoryRequest)
+
+	return replicas, &resources, nil
+}
+
+func constructSMFNadName(templateName string, suffix string) string {
+	return templateName + "-" + suffix
+	// return "n4"
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:105]: remove this line

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+// getSMFNad returns NAD label string composed based on the Nx interfaces configuration provided in SMFDeploymentSpec
+func getSMFNad(templateName string, spec *workloadv1alpha1.SMFDeploymentSpec) string {
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:112]: are these commented out things not needed or is this just temporary and will be updated in later PRs?

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+		// "n9": n9CfgSlice,
+	}
+	// Need to sort inftMap by key otherwise unitTests might fail as order of intefaces in intfMap is not guaranteed
+	inftMapKeys := make([]string, 0, len(intfMap))
+	for interfaceName := range intfMap {
+		inftMapKeys = append(inftMapKeys, interfaceName)
+	}
+	sort.Strings(inftMapKeys)
+
+	noComma := true
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:131]: is this JSON we're producing here for something to consume? There's probably a more idiomatic way to do it that maps structs to JSON. That is, you create Go structs for the things you need, with JSON tags. In fact if these are Multus things they probably already have such types. So, you would import that module, construct the objects in Go, and then use json.Marshal. I *think* that would handle the sorting of the map keys too, but not sure.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+}
+
+// checkNADExists gets deployment object and checks "k8s.v1.cni.cncf.io/networks" NADs.
+// returns True if all requred NADs are present
+// returns False if any NAD doesn't exists in deployment namespace
+func (r *SMFDeploymentReconciler) checkSMFNADexist(log logr.Logger, ctx context.Context, deployment *appsv1.Deployment) bool {
+	smfAnnotations := []SMFAnnotation{}
+	annotationsString, ok := deployment.Spec.Template.GetAnnotations()["k8s.v1.cni.cncf.io/networks"]
+	if !ok {
+		log.Info("Annotations k8s.v1.cni.cncf.io/networks not found", "SMFDeployment.namespace", deployment.Namespace)
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:167]: Do we need to log this? Seems too verbose. Is there a Debug log level?

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+		return false
+	}
+	if err := json.Unmarshal([]byte(annotationsString), &smfAnnotations); err != nil {
+		log.Info("Failed to parse SMFDeployment annotations", "SMFDeployment.namespace", deployment.Namespace)
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:171]: So, this is an error, not a false return. The method should return (bool, error) and in this case return the error.

We shouldn't need to log here either - even if there is an error. I would expect the caller to handle the error (which may include logging or something else).

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 16]
```diff
@@ -0,0 +1,648 @@
...
+	for _, smfAnnotation := range smfAnnotations {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   "k8s.cni.cncf.io",
+			Kind:    "NetworkAttachmentDefinition",
+			Version: "v1",
+		})
+		key := client.ObjectKey{Namespace: deployment.ObjectMeta.Namespace, Name: smfAnnotation.Name}
+		if err := r.Get(ctx, key, u); err != nil {
+			log.Info(fmt.Sprintf("Failed to get NAD %s", smfAnnotation.Name), "SMFDeployment.namespace", deployment.Namespace)
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:183]: So, yeah, we need to change the method to return (bool, error). In this case, if the error is "not found", I think that's not an error but part of what the method is checking, correct? In that case, we would `return false, nil` and be done. But if it's a different error, then it indicates some failure and we should `return false, err` and let the caller figure out what to do about the error (most likely, it will be to set a failed Ready condition in the object status and surface the error to the user).

Here's an example of using the `IgnoreNotFound` - though you should look up exactly what it returns as I am not sure offhand: https://github.com/GoogleContainerTools/kpt/blob/main/porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go#L147

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 17]
```diff
@@ -0,0 +1,648 @@
...
+		}
+	}
+
+	return true
+}
+
+func free5gcSMFDeployment(log logr.Logger, smfDeploy *workloadv1alpha1.SMFDeployment) (*appsv1.Deployment, error) {
+	//TODO(jbelamaric): Update to use ImageConfig spec.ImagePaths["smf"],
+	// smfImage := "nephio/free5gc-smf:latest"
+	smfImage := "towards5gs/free5gc-smf:v3.2.0"
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:194]: We probably need to be able to read in, at the least, the registry and the tag. But I think this is OK for now (though we have our own built images in the nephio docker now).

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+// Check for UERoute in SMFConfigMap yaml
+// https://github.com/s3wong/free5gc/blob/main/doc/sample-manifests/smf.yaml#L122
+// https://github.com/s3wong/free5gc/blob/main/doc/sample-manifests/smf.yaml#L272
+func free5gcSMFCreateConfigmap(logger logr.Logger, smfDeploy *workloadv1alpha1.SMFDeployment) (*apiv1.ConfigMap, error) {
+	namespace := smfDeploy.ObjectMeta.Namespace
+	instanceName := smfDeploy.ObjectMeta.Name
+
+	n4IP, err := getIPv4(smfDeploy.Spec.Interfaces, "n4")
+	if err != nil {
+		log.Log.Info("Interface N4 not found in NFDeployment Spec")
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:302]: No need to log here. For error handling, I suggest we log at the highest possible level, and just return the error at the lower levels. We can add context if necessary with a new error wrapping the underlying one, though generally I suggest we avoid that unless absolutely necessary.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 19]
```diff
@@ -0,0 +1,648 @@
...
+		return nil, err
+	}
+
+	smfcfgStruct := SMFcfgStruct{}
+	smfcfgStruct.PFCP_IP = n4IP
+
+	/*
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:309]: Can we delete this comment?

If we want an example of what the config file looks like, perhaps we can include that. But we don't need the K8s yaml pieces, at the least.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 20]
```diff
@@ -0,0 +1,648 @@
...
+		        ReportCaller: false
+		        debugLevel: info
+		      SMF:
+		        ReportCaller: false
+		        debugLevel: info
+	*/
+	smfcfgTemplate := template.New("SMFCfg")
+	smfcfgTemplate, err = smfcfgTemplate.Parse(SMFCfgTemplate)
+	if err != nil {
+		log.Log.Info("Could not parse SMFCfgTemplate template.")
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:419]: for all these don't log, just wrap the error if you need to add the context of which template you failed to parse. See https://go.dev/blog/go1.13-errors for a discussion, in particular the "Wrapping errors with %w" section.

TL;DR, instead of log and return, just do:
```go
return nil, fmt.Errorf("SMFCfgTemplate parse error: %w", err)
```

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 21]
```diff
@@ -0,0 +1,648 @@
...
+		newSmf := smfDeploy
+		newSmf.Status.NFDeploymentStatus = newSMFStatus
+		err := r.Status().Update(ctx, newSmf)
+		return err
+	}
+
+	return nil
+}
+
+func calculateSMFStatus(deployment *appsv1.Deployment, smfDeploy *workloadv1alpha1.SMFDeployment) (workloadv1alpha1.NFDeploymentStatus, bool) {
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:473]: We should have a `Ready` condition. Any error that occurs during processing should set the `Ready` condition to `False` and set the `Message` to the error string. This is how we surface errors to the user.

Well, there are basically two ways to surface errors. We should never expect the user to see a log file. Those are just for us, really. The resource status conditions is the primary way to surface an error *state*. However you can also fire *events*. Events in Kubernetes *are NOT for machine consumption*. They are only for users. They will show up, for example, when you `kubectl describe` the resource.

Currently I don't think we're firing events almost anywhere in Nephio. But eventually we should.

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 22]
```diff
@@ -0,0 +1,648 @@
...
+		condition.Message = "SMFDeployment pods are available."
+		condition.LastTransitionTime = metav1.Now()
+	} else if lastDeploymentStatus.Type == appsv1.DeploymentProgressing {
+		condition.Type = string(workloadv1alpha1.Reconciling)
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = "MinimumReplicasNotAvailable"
+		condition.Message = "SMFDeployment pod(s) is(are) starting."
+		condition.LastTransitionTime = metav1.Now()
+	} else if lastDeploymentStatus.Type == appsv1.DeploymentReplicaFailure {
+		condition.Type = string(workloadv1alpha1.Stalled)
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:524]: I think these condition types are probably OK. FYI: https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus#conditions

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 23]
```diff
@@ -0,0 +1,648 @@
...
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *SMFDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx).WithValues("SMFDeployment", req.NamespacedName)
+
+	smfDeploy := &workloadv1alpha1.SMFDeployment{}
+	err := r.Client.Get(ctx, req.NamespacedName, smfDeploy)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("SMFDeployment resource not found. Ignoring since object must be deleted")
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:562]: Probably don't need to log

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 24]
```diff
@@ -0,0 +1,648 @@
...
+
+	namespace := smfDeploy.ObjectMeta.Namespace
+	// see if we are dealing with create or update
+	cmFound := false
+	configmapName := smfDeploy.ObjectMeta.Name + "-smf-configmap"
+	currConfigmap := &apiv1.ConfigMap{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: configmapName, Namespace: namespace}, currConfigmap); err == nil {
+		cmFound = true
+	}
+
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:577]: are we swallowing errors? Or are they not errors but just means we have to create not update?

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 25]
```diff
@@ -0,0 +1,648 @@
...
+			if err := r.syncSMFStatus(ctx, d, smfDeploy); err != nil {
+				log.Error(err, "Failed to update SMFDeployment status", "SMFDeployment.namespace", namespace, "sMFDeployment.name", smfDeploy.Name)
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// first set up the configmap
+	if cm, err := free5gcSMFCreateConfigmap(log, smfDeploy); err != nil {
+		log.Error(err, fmt.Sprintf("Error: failed to generate configmap %s\n", err.Error()))
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:600]: We should be surfacing this in the Status

File: free5gc-operator/controllers/smfdeployment_controller.go [hunk 26]
```diff
@@ -0,0 +1,648 @@
...
+			log.Info("Creating SMFDeployment configmap", "SMFDeployment.namespace", namespace, "Confirmap.name", cm.ObjectMeta.Name)
+			// Set the controller reference, specifying that SMFDeployment controling underlying deployment
+			if err := ctrl.SetControllerReference(smfDeploy, cm, r.Scheme); err != nil {
+				log.Error(err, "Got error while setting Owner reference on configmap.", "SMFDeployment.namespace", namespace)
+			}
+			if err := r.Client.Create(ctx, cm); err != nil {
+				log.Error(err, fmt.Sprintf("Error: failed to create configmap %s\n", err.Error()))
+				return reconcile.Result{}, err
+			}
+		}
```
Comment (johnbelamaric) [inline free5gc-operator/controllers/smfdeployment_controller.go:613]: Similar for above errors, they need to be surfaced in the status

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Some comments. I think the error handling needs some work.

Also, I ran out of time so maybe I just missed it, but I don't see a finalizer being set? This means if someone deletes the SMFDeployment object, we will orphan all the resources we created. We should be tagging all of the resources we create with ownerReferences, and add a finalizer to our resource. When someone deletes the resource, it sets the DeletionTimestamp, and so when we see that in the resource during reconciliation, we clean up all our owned resources (using the owner refs), then clear our finalizer.

I would be OK merging this and then adding all these fixes in follow up PRs, so we can start e2e testing. But if we do that, please create issues for the different things that need fixing.

**Comment by johnbelamaric** [comment]
@n2vo great start! I added some comments and some thoughts on how to improve things.

File: free5gc-operator/controllers/smfdeployment_controller.go (hunk 21 above)
Comment (johnbelamaric) [reply free5gc-operator/controllers/smfdeployment_controller.go:473]: Note that that link does NOT recommend Ready condition, but I think they are wrong.

File: free5gc-operator/controllers/smfdeployment_controller.go (hunk 17 above)
Comment (n2vo) [reply free5gc-operator/controllers/smfdeployment_controller.go:194]: > We probably need to be able to read in, at the least, the registry and the tag. But I think this is OK for now (though we have our own built images in the nephio docker now).
Agreed, @johnbelamaric , for now it's hardcoded but will add code for reading in the image/version.  As for the docker.io/nephio. I tried referencing that but didn't have access.  Working thru that with @s3wong and SIG Release now.

File: free5gc-operator/Dockerfile (hunk 1 above)
Comment (n2vo) [reply free5gc-operator/Dockerfile]: Done

**Comment by denysaleksandrov** [comment]
@johnbelamaric regarding Finalizers. I initially added Finalizers to UPFDeployment, which is used as a base for AMF/SMFDeployment code. Then @tliron suggested against using finalizers and I refactored my code to use ControllerReference:
```go
...
			// Set the controller reference, specifying that UPFDeployment controls underlying deployment
			if err := ctrl.SetControllerReference(upfDeploy, cm, r.Scheme); err != nil {
				log.Error(err, "Got error while setting Owner reference on configmap.", "UPFDeployment.namespace", namespace)
			}
...
				// Set the controller reference, specifying that UPFDeployment controls underlying deployment
				if err := ctrl.SetControllerReference(upfDeploy, deployment, r.Scheme); err != nil {
					log.Error(err, "Got error while setting Owner reference on deployment.", "UPFDeployment.namespace", namespace)
				}
```
So that allows to controller to delete both dependency config-map and deployment objects if UPFDeployment is deleted.
The same is implemented in the SMF controller. We discussed it many times in sig-auto-workload meetings.
I tested it, if UPFDeployment is deleted, all dependency objects are deleted as well.

**Comment by johnbelamaric** [comment]
@denysaleksandrov Ah, right. The finalizers are only needed when dealing with resources external to the cluster. API server will garbage collect based on owner refs. That's great, thanks.

**Review by tliron — APPROVED** [review]
